### PR TITLE
Add a warning if spec is applied on a dirty worktree

### LINKED ISSUE #1181: fission spec apply -- warn or error on dirty git tree
Since fission spec apply is often used for applying specs from a git repo, it would be useful for it to warn or refuse to work (by defautl) when the git tree is dirty.

### PR DESCRIPTION
**Pull request by hellozee**
Fixes #1181

### CODE REVIEW

**Comment by vishal-biyani** [comment]
Hey @hellozee Thanks for this change - very useful.

Can you please paste some screen/text grabs of testing done on CLI and resolve conflict and we are good to merge!

**Comment by hellozee** [comment]
Here is a screenshot, 
ছবি

Also, I am not sure about the updates to go.sum/go.mod files.

File: go.mod [hunk 1]
```diff
@@ -1,5 +1,7 @@
 module github.com/fission/fission
 
+go 1.15
```
Comment (vishal-biyani) [inline go.mod:3]: We will have to switch to version that is in use right now - for now I think is 1.12.X - can you please change that?

**Comment by vishal-biyani** [comment]
One last request @hellozee Can you please help with conflicts?

File: pkg/fission-cli/cmd/spec/apply.go [hunk 2]
```diff
@@ -185,6 +188,27 @@ func (opts *ApplySubCommand) run(input cli.Input) error {
 	return nil
 }
 
+func warnIfDirtyWorkTree(path string) {
+	repo, err := git.PlainOpen(path)
+	if err != nil {
+		return
```
Comment (sanketsudake) [inline pkg/fission-cli/cmd/spec/apply.go:194]: The function should not ignore error silently. Can you please return an error from all places, let the caller decide how to handle the error.

File: pkg/fission-cli/cmd/spec/apply.go [hunk 3]
```diff
@@ -123,6 +124,8 @@ func (opts *ApplySubCommand) run(input cli.Input) error {
 			}
 		}
 
+		warnIfDirtyWorkTree(filepath.Clean(specDir + "/.."))
```
Comment (sanketsudake) [inline pkg/fission-cli/cmd/spec/apply.go:127]: If this function returns an error we should log it here, but not fail the entire program. This helps us in making sure this function is working properly from a future maintenance perspective.

File: pkg/fission-cli/cmd/spec/apply.go [hunk 4]
```diff
@@ -185,6 +188,27 @@ func (opts *ApplySubCommand) run(input cli.Input) error {
...
+		return
+	}
+
+	status, err := workTree.Status()
+	if err != nil {
+		return
+	}
+
+	if !status.IsClean() {
+		console.Warn("Worktree is not clean")
```
Comment (sanketsudake) [inline pkg/fission-cli/cmd/spec/apply.go:208]: We are giving a warning to the user, we should also provide an action to resolve them.
```
worktree is not clean, please ensure you have committed changes to git.
```

File: pkg/fission-cli/cmd/spec/apply.go [hunk 5]
```diff
@@ -185,6 +188,27 @@ func (opts *ApplySubCommand) run(input cli.Input) error {
 	return nil
 }
 
+func warnIfDirtyWorkTree(path string) {
+	repo, err := git.PlainOpen(path)
```
Comment (sanketsudake) [inline pkg/fission-cli/cmd/spec/apply.go:192]: Would this work if a user is not using Git in his project or it's just a local folder and not git repo. Can we check if they have enabled Git or not in their project.

Comment (hellozee) [reply pkg/fission-cli/cmd/spec/apply.go:192]: It would return an error which I ignore and return from that point, as you pointed above

Comment (sanketsudake) [reply pkg/fission-cli/cmd/spec/apply.go:192]: As per documentation 
https://pkg.go.dev/gopkg.in/src-d/go-git.v4#PlainOpen
PlainOpen calls returns ErrRepositoryNotExists if not valid repository, we should ignore this error only. Rest we should log.

Comment (hellozee) [reply pkg/fission-cli/cmd/spec/apply.go:192]: Okay

**Review by sanketsudake — APPROVED** [review]

**Comment by sanketsudake** [comment]
Thanks for you contribution @hellozee 🎉

**Comment by sjk7524068** [comment]
> 

I met the same problem, what dose this mean, should I do something to fix ?

**Pull request merged**